### PR TITLE
fix(mcdu): performance page refresh on flight phase transition

### LINKED ISSUE #4635: (MCDU) All PERF pages are still accessible at a particular situation during approach
**Aircraft Version**

```json
{
  "built": "2021-04-28T03:09:12+00:00",
  "ref": "master",
  "sha": "ac38d5908a21967d040bacfe97b31d76f3ec202d",
  "actor": "Benjozork",
  "event_name": "manual"
}

```

**Describe the bug**

Looks like something that was fixed in https://github.com/flybywiresim/a32nx/pull/2940

There is still a way to access all the previous MCDU PERF pages such as climb, takeoff, etc. When on descent, scroll to the APPR page, fill the details such as TRANS ALT, QNH, TEMP, etc. and continue descent to an extent where you cross the DECEL point, and/or are below 9500ft (approx., as per my test airport VOML), basically **until the Approach phase gets automatically engaged on the MCDU,** without any intervention such as clicking the confirm activate APPR button, etc.

Once the MCDU approach phase is engaged, try using the "PREV PHASE" option (which is still visible , while it should have disappeared right after APPR is engaged [possible code hint] ) to scroll back through the earlier phases.

https://cdn.discordapp.com/attachments/754130199804772372/836987727135637504/Desktop_2021.04.28_-_20.59.33.03.mp4

**To Reproduce**

1. Descend normally, with APPR page opened on the MCDU, and all the details filled up
2. Do not engage MCDU Approach Phase manually on DES page, fly through the decel point, and continue on the approach with the APPR page open
3. Click the PREV PHASE option and observe older PERF pages showing up

**Expected behavior**

When the MCDU approach phase is engaged automatically while the MCDU displays the APPR page, an update code needs to run to hide the "PREV PHASE" option, or simply put, update the status of the page.

**Actual behavior**

Even though the Approach phase was engaged automatically, the PREV PHASE option still remained on display, allowing access to all the previous phases

**References**

**Additional context**

Was this working before/when did the issue start occurring?
This is a very rare situation, may have existed all this time, but went unnoticed

Is this a problem in the vanilla unmodded game?
No

Discord username (if different from GitHub): sidnov#8337

### PR DESCRIPTION
**Pull request by MisterChocker**
Fixes #4635

## Summary of Changes

- Added additional check for current performance page and current flight phase
- Fixed page refresh when when having next flight phases performance page open before flight phase transition into next phase
e.g. flight phase descent, performance page approach -> flight phase transition into approach -> performance page didn't reload (now reloads)

## Additional context

Discord username (if different from GitHub): DerL30N#3751

## Testing instructions

Any flight phase transition observe performance page is updated on every phase change.

Best example to test with:
From phase DES into APPR.
Go to performance page APPR during DES phase.
When reaching DECEL waypoint and FMGC transitions to flight phase approach -> monitor performance page reload (prev page indication and action will disappear)

Use managed speed mode and monitor pfd managed speed target to ensure approach phase is active (speed target GD speed)
 

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js [hunk 1]
```diff
@@ -374,7 +374,7 @@ class FMCMainDisplay extends BaseAirliners {
             }
 
             case FmgcFlightPhases.CRUISE: {
-                if (this.canShowNextPerfPage(_lastFlightPhase)) {
+                if (this.canShowNextPerfPage(_lastFlightPhase) || this.canShowNextPerfPage(_curFlightPhase)) {
```
Comment (beheh) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:377]: Are we sure about this? If one MCDU re-establishes CRZ from APPR (by re-entering a CRZ flight level) are we sure the other one should leave APPR to go to CRZ? APPR is still perfectly valid, as it's a future flight phase.

Comment (MisterChocker) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:377]: @beheh this has nothing todo with that, it only ensures to switch to the next flight phases page in order or reload the the one for the current page if already selected.

Comment (beheh) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:377]: > it only ensures to switch to the next flight phases page in order or reload the the one for the current page if already selected.

No, that's exactly what I mean though. Doesn't this line mean that if your MCDU has APPR open, and you reengage CRZ through some other way (i.e. the other MCDU), we switch the page back to CRZ? This switch generally makes sense when going "forwards" through the flight phases, for example of course when we have the TAKE OFF page open and we go to CLB, we want to switch the page to CLB (TAKE OFF even becomes unreachable). But if we have APPR open, which is *after* CRZ, and somehow CRZ reengages, should the really page go "backwards" to reach CRZ? It seems perfectly valid to have APPR open during either CRZ and APPR, so I'm not sure it should automatically go backwards.

Comment (MisterChocker) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:377]: I still don't understand the issue, when you even if you change flight phases in an oddly order why shouldn't the performance page allow you to go to that flight phase again?

Either way, when in APPR and going into CRZ, the page won't automatically transition onto that page.
It only goes to that page when the page was a) the page one before, or b) the page was already open and is the corresponding page to the new flight phase.

Basically:
- CLB -> CRZ (while on perf CLB page): perf page auto transition;
 - CLB -> CRZ (while on perf CRZ page)): crz perf page auto reload;
 - CLB -> CRZ (while on perf DES page): no autotransition or auto reload;
 - APPR -> CRZ (while on perf APPR page): perf page auto transition:
 - APPR -> CRZ (while on GO AROUND page): no autotransition or auto reload;

Comment (beheh) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:377]: > APPR -> CRZ (while on perf APPR page): perf page auto transition:

This is the behavior I'm questioning, and whether you're sure about it. The autotransition going forward makes sense, because the old page becomes unreachable, however when you transition to CRZ, APPR stays continuously available. Why should it auto-transition away from APPR in this case?

Comment (MisterChocker) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:377]: Now I understand, I assumed that the performance page (when displaying the current page relative to flight phase) will transition to the page corresponding to the new flight phase.

Comment (MisterChocker) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:377]: I'll have to check if I can find something about this behavior in our references, but I highly doubt that.

Comment (MisterChocker) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:377]: okay I wasn't able to find anything, BUT it's impossible for the fmgc to automatically transition from APPROACH to a previous flight phase, which means the issue is no longer relevant as it only covers automatic flight phase transitions.

Manual transitions from APPR to CLB or CRZ have to be made via other MCDU pages and therefore an automatic PERF PAGE change is impossible.

Comment (MisterChocker) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js]: @beheh I've improved and updated the code to follow your suggestion.
Only "autochange" performance pages "forward".

**Review by beheh — APPROVED** [review]

**Review by sidnov — APPROVED** [review]

**Comment by MisterChocker** [comment]
Found an issue in the code, need to fix it first.

**Comment by HPringles** [comment]
Quality Assurance Trainee Report

Object of testing: #4646 
Tier of Testing: 1
Date: 24/05/2021

Testing Process: 

- Full Flight - EGKK to EGSS
- Before each phase started, switched to the next page and monitored the phase change
- Once the next phase started, the page would update, most notably to remove the option to return to the previous stage
- 

Negatives: N/A
Testing Results: Passed
Conclusions: Working as expected - LGTM

**Comment by DavidEGTE** [comment]
Quality Assurance Trainee Report

Discord: David#2177
Object of testing: #4646 
Tier of Testing: 1
Date: 24/05/2021

Testing Process: 
- Full Flight - LIRN to LICC
- Monitored every single phase change as per testing instructions

Negatives: N/A
Testing Results: Passed
Conclusions: Working as expected

**Comment by ExampleWasTaken** [comment]
Quality Assurance Tester

Object of testing: #4646 
Tier of Testing: 1
Date: 25/05/2021

Testing Process: 
1. Loaded C&D
2. Powered up using normal procedure
3. Took of
4. Climbed to 10000 ft
5. Initiated descent 
6. Landed and taxied to stand

I monitored every flight phase change during the flight. See issues.


Negatives: Passing the DECEL point while in DEC phase "Activate APPR" was removed and replaced by "PREV PHASE". Also, the title of the DES page turned white. It did not switch to APPR page.

Testing Results: Not Passed

Conclusions: Other than the mentioned issue everything worked as expected. 👍

**Comment by MisterChocker** [comment]
Updated the branch.

@ExampleWasTaken I'am looking into it.

**Comment by TheSquirtingPantin** [comment]
Category: QA Tester
Name: Discord User Rémi#8871
Date of testing: Current date 26/05/2021
Version of the sim: V 1.16.2.0
PR Tested: #4646
Tier of testing: 2
Changes to observe: PERF page adjustment and logic

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. C/D at LFBO
2. Full Flight
3. Observe behavior of PERF page


Flight notes (When done a full flight): LFBO/32L DCT LACOU UT24 SECHE UL127 SAU DCT SOMOS DCT LMB DCT LFBO/32L

Media:

Issues (If Any):
1) During cruise, opened the PERF page, then transitioned to an OP DES. PERF page did not transitioned to DES. Clicked NEXT PAGE and no further issues displayed (CRZ page no longer available, DES page correct)
2) No issues with transition to APPR phase post DECEL. 

Overall Rating: To be Improved
Conclusions of the testing: Issue with transition from CRZ to DES phase.

**Review by tracernz — APPROVED** [review]

**Comment by AviationSpotter** [comment]
Quality Assurance Trainee Report

Discord: Aviation Spotter#9102
Object of testing: #4646
Tier of Testing: 1
Date: 10/6/2021

Testing Process:

1. Spawned C&D
2. Full Flight (KTPA -> KFLL)
3. Performance page was monitored through each flight phase transition

Negatives: N/A
Testing Results: Passed

Conclusions: During the flight, the flight phase transitions behaved as intended.

**Comment by marcman86** [comment]
Quality Assurance Tester Report
Discord:marcman86#4907
Object of testing: #4646
Tier of Testing: 2
Date: 11/06/2021

Testing Process: 
- Full flight from EDDM to LIPE
- Checked Perf Page the whole time

Negatives: -
Testing Results: Passed
Conclusions: Works
Media: -
Hope this helps :-)
